src/main/java/au/org/ala/biocache/util/thread/DownloadControlThread.java

```diff
@@ -0,0 +1,141 @@
+import au.org.ala.biocache.dto.DownloadDetailsDTO;
+import au.org.ala.biocache.dto.DownloadDetailsDTO.DownloadType;
+
+/**
+ * A Runnable that can be used to schedule records dumps for a particular class of download.
+ * 
+ * @author Peter Ansell
+ */
+public class DownloadControlThread implements Runnable {
+    private final Integer maxRecords;
+    private final DownloadType downloadType;
+    private final int concurrencyLevel;
+    private final long pollDelay;
+    private final long executionDelay;
+    private final int threadPriority;
+    
+    private final DownloadServiceExecutor downloadServiceExecutor;
+    private final AtomicBoolean shutdownFlag = new AtomicBoolean(false);
+    private final Queue<DownloadDetailsDTO> currentDownloads;
+    private final DownloadCreator downloadCreator;
+    private final PersistentQueueDAO persistentQueueDAO;
+    
+    public DownloadControlThread(Integer maxRecords, DownloadType downloadType, int concurrencyLevel, Long pollDelayMs, Long executionDelayMs, Integer threadPriority, Queue<DownloadDetailsDTO> currentDownloads, DownloadCreator downloadCreator, PersistentQueueDAO persistentQueueDAO) {
+        this.maxRecords = maxRecords;
+        this.downloadType = downloadType;
+        this.concurrencyLevel = concurrencyLevel > 0 ? concurrencyLevel : 1;
+        this.pollDelay = pollDelayMs != null && pollDelayMs >= 0L ? pollDelayMs : 10L;
+        this.executionDelay = executionDelayMs != null && executionDelayMs >= 0L ? executionDelayMs : 0L;
+        this.threadPriority = threadPriority != null && threadPriority >= Thread.MIN_PRIORITY && threadPriority <= Thread.MAX_PRIORITY ? threadPriority : Thread.NORM_PRIORITY;
+        this.currentDownloads = currentDownloads;
+        this.downloadCreator = downloadCreator;
+        this.persistentQueueDAO = persistentQueueDAO;
+        // Create a dedicated ExecutorService for this thread
+        this.downloadServiceExecutor = createExecutor();
+    }
+
+    protected DownloadServiceExecutor createExecutor() {
+        return new DownloadServiceExecutor(this.maxRecords, this.downloadType, 
+                                           this.concurrencyLevel, this.executionDelay, 
+                                           this.threadPriority, this.downloadCreator);
+    }
+
+    @Override
+    public void run() {
+        DownloadDetailsDTO currentDownload = null;
+        try {
+            // Runs until the thread is interrupted, then shuts down all of the downloads under its control before returning
+            while (true) {
+                if(shutdownFlag.get() || Thread.currentThread().isInterrupted()) {
+                    break;
+                }
+                // Busy wait polling
+                // TODO: Convert PersistentQueueDAO to a
+                // blocking/interruptible wait to avoid busy wait
+                Thread.sleep(pollDelay);
+                currentDownload = persistentQueueDAO.getNextDownload(maxRecords, downloadType);
```

Comment:
I redesigned this code to use a Semaphore as a capacity wrapper here, which also makes it possible to remove the Thread.sleep(pollDelay) and replace it with a call to Semaphore.tryAcquire(). This will change the semantics a little during startup, as all of the executors will fill up eagerly with whatever downloads they are eligible for, but in the long term during execution it will have similar semantics. The no-limit download threads may pickup some smaller downloads during this time, but that shouldn't affect them in the long term. To work around that with this design we would need a minimum-limit added to download threads, but it shouldn't be an actual issue.